akka-testkit/src/main/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala

```diff
@@ -0,0 +1,107 @@
+package akka.testkit
+
+import java.util.concurrent.ThreadFactory
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicLong
+
+import com.typesafe.config.Config
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.{ Duration, FiniteDuration }
+import akka.actor.Cancellable
+import akka.actor.Scheduler
+import akka.event.LoggingAdapter
+
+/**
+ * For testing: scheduler that does not look at the clock, but must be
+ * progressed manually by calling `timePasses`.
+ *
+ * This allows for faster and less timing-sensitive specs, as jobs will be
+ * executed on the test thread instead of using the original
+ * {ExecutionContext}. This means recreating specific scenario's becomes
+ * easier, but these tests might fail to catch race conditions that only
+ * happen when tasks are scheduled in parallel in 'real time'.
+ */
+class ExplicitlyTriggeredScheduler(config: Config, log: LoggingAdapter, tf: ThreadFactory) extends Scheduler {
+
+  case class Item(time: Long, interval: Option[FiniteDuration], runnable: Runnable)
+
+  val currentTime = new AtomicLong()
+  val scheduled = new ConcurrentHashMap[Item, Unit]()
+
+  override def schedule(initialDelay: FiniteDuration, interval: FiniteDuration, runnable: Runnable)(implicit executor: ExecutionContext): Cancellable =
+    schedule(initialDelay, Some(interval), runnable)
+
+  override def scheduleOnce(delay: FiniteDuration, runnable: Runnable)(implicit executor: ExecutionContext): Cancellable =
+    schedule(delay, None, runnable)
+
+  /**
+   * Advance the clock by the specified duration, executing all outstanding jobs on the calling thread before returning.
+   *
+   * We will not add a dilation factor to this amount, since the scheduler API also does not apply dilation.
+   * If you want the amount of time passed to be dilated, apply the dilation before passing the delay to
+   * this method.
+   */
+  def timePasses(amount: FiniteDuration) = {
+    // Give dispatchers time to clear :(
```

Comment:
when is that needed and what will happen if they are not done within 10 ms?

Reply:
It is needed when:
* the test triggers some asynchronous operations
* those asynchronous operations schedule tasks
* immediately after triggering the asynchronous operations, we manually advance the scheduler to a point in time after the deadline for the scheduled tasks

This is a race condition: when the scheduler queue is checked before the tasks are scheduled, they are (obviously) missed and won't be evaluated.

The 'proper' solution might be to introduce methods like `expectMsg` and friends that are aware of the manually triggered scheduler and can retry checking the task queue before failing. That would mean we re-use less of the `TestProbe` implementation, perhaps we should even have  separate `TestProbe` implementations, for use with 'normal' schedulers and one for use with the explicit scheduler...

Reply:
Alternatively, we could consider taking control over the dispatcher as well (perhaps for now just allowing inspection), so we can observe whether it is finished. Could be interesting but would also mean a bit of extra code - WDYT?

Reply:
I'm not quite sure what kind of rabbit whole this is. I'll probably prefer something simple until we get a feeling of what kind of problems/limitations there are. Perhaps @jrudolph has any thoughts.

Reply:
> Alternatively, we could consider taking control over the dispatcher as well

That's something that I would have found useful quite a few times as well. The question is whether and how we could reliably determine that a dispatcher is done for now? We could say it's good enough if the task queue is empty and no threads are running (ignoring external task submissions). But even that might be hard to check atomically because executors tend to be optimized not to need a single lock.

I guess we could define a dispatcher wrapper for testing that would keep track of all task submission before submitting to the real dispatcher counting an atomic long up and down?